spec/datadog/core/remote/component_spec.rb

```diff
@@ -0,0 +1,214 @@
+      expect(worker).to receive(:stop).and_call_original
+    end
+
+    after do
+      component.shutdown!
+    end
+
+    context 'when client sync succeeds' do
+      before do
+        expect(worker).to receive(:call).and_call_original
+        expect(client).to receive(:sync).and_return(nil)
+      end
+
+      it 'does not log any error' do
+        expect(Datadog.logger).to_not receive(:error)
+
+        component.barrier(:once)
+      end
+    end
+
+    context 'when client sync raises' do
+      let(:second_client) { double }
+      let(:exception) { Class.new(StandardError) }
+
+      before do
+        expect(worker).to receive(:call).and_call_original
+        expect(client).to receive(:sync).and_raise(exception, 'test')
+        allow(Datadog.logger).to receive(:error).and_return(nil)
+      end
+
+      it 'logs an error' do
+        allow(Datadog::Core::Remote::Client).to receive(:new).and_return(client)
+
+        expect(Datadog.logger).to receive(:error).and_return(nil)
+
+        component.barrier(:once)
+      end
+
+      it 'catches exceptions' do
+        allow(Datadog::Core::Remote::Client).to receive(:new).and_return(client)
+
+        # if the error is uncaught it will crash the test, so a mere passing is good
+
+        component.barrier(:once)
+      end
+
+      it 'creates a new client' do
+        expect(Datadog::Core::Remote::Client).to receive(:new).and_return(second_client)
+
+        expect(component.client.object_id).to eql(client.object_id)
+
+        component.barrier(:once)
+
+        expect(component.client.object_id).to eql(second_client.object_id)
+      end
+    end
+  end
+end
+
+RSpec.describe Datadog::Core::Remote::Component::Barrier do
+  let(:delay) { 0.5 }
+  let(:block) { -> { true } }
+
+  subject(:barrier) { described_class.new(&block) }
+
+  shared_context('recorder') do
+    let(:record) { [] }
+  end
+
+  shared_context('waiter thread') do
+    include_context 'recorder'
+
+    let(:thr) do
+      Thread.new do
+        loop do
+          record << :wait
+          barrier.wait_next
+        end
+      end
+    end
+
+    before do
+      record
+      thr.run
+    end
+
+    after do
+      thr.kill
+      thr.join
+    end
+  end
+
+  shared_context('lifter thread') do
+    include_context 'recorder'
+
+    let(:thr) do
+      Thread.new do
+        loop do
+          sleep delay
+          record << :lift
+          barrier.lift
+        end
+      end
+    end
+
+    before do
+      record
+      thr.run
+    end
+
+    after do
+      thr.kill
+      thr.join
+    end
+  end
+
+  describe '#lift' do
+    context 'without waiters' do
+      include_context 'recorder'
+
+      it 'does not block' do
+        record << :one
+        barrier.lift
+        record << :two
+
+        expect(record).to eq [:one, :two]
+      end
+    end
+
+    context 'with waiters' do
+      include_context 'waiter thread'
+
+      it 'unblocks waiters' do
+        sleep delay
```

Comment:
@lloeki, we talked about it. Just leaving it here for reference.  It would be good to try using `Queue` rather than sleep. Since the intention of the sleep if to force the VM to switch to another thread, we could use `queue.pop` to wait until the other thread unblock us using `queue << 1`

Reply:
That is a very fair point. 

Out of curiosity, I tried to make it work with `Queue`

Here is the diff, the tests are passing 
![image](https://user-images.githubusercontent.com/4672858/231983363-2bf88e95-b1f7-4fe4-9a03-376b95847d25.png)

Reply:
I am slightly worried that test correctness here is dependent on racy sleeps as well.